tests/unit_tests/oci/test_instance.py

```diff
@@ -0,0 +1,475 @@
+            ),
+            pytest.param(
+                False,  # primary_has_ipv4
+                True,  # primary_has_ipv6
+                True,  # secondary_has_ipv4
+                False,  # secondary_has_ipv6
+                False,  # primary_first
+                "2001:db8::1",  # expected_ip
+                False,  # expect_error
+                id="Multiple primary VNICs, primary not first (IPv6)",
+            ),
+            pytest.param(
+                True,  # primary_has_ipv4
+                False,  # primary_has_ipv6
+                False,  # secondary_has_ipv4
+                False,  # secondary_has_ipv6
+                False,  # primary_first
+                "203.0.113.1",  # expected_ip
+                False,  # expect_error
+                id="Primary VNIC has IPv4 (primary not first)",
+            ),
+        ],
+    )
+    def test_oci_instance_ip_parametrized(
+        self,
+        oci_instance,
+        primary_has_ipv4,
+        primary_has_ipv6,
+        secondary_has_ipv4,
+        secondary_has_ipv6,
+        primary_first,
+        expected_ip,
+        expect_error,
+    ):
+        """
+        Test the ip property of OciInstance with various VNIC configurations.
+
+        By testing a matrix of VNIC configurations, this test ensures that the ip property
+        of OciInstance behaves as expected in various scenarios, testing all meaningful
+        combinations of the following conditions:
+        - if the primary VNIC has an IPv4 address
+        - if the primary VNIC has an IPv6 address
+        - if the secondary VNIC has an IPv4 address
+        - if the secondary VNIC has an IPv6 address
+        - if the primary VNIC is listed first
+        """
+        # Prepare VNIC configurations
+        # Primary VNIC configuration
+        primary_vnic = (
+            "1",  # vnic_id_suffix
+            True,  # is_primary
+            "203.0.113.1" if primary_has_ipv4 else None,  # public_ip
+            ["2001:db8::1"] if primary_has_ipv6 else [],  # ipv6_addresses
+        )
+
+        # Secondary VNIC configuration
+        secondary_vnic = (
+            "2",  # vnic_id_suffix
+            False,  # is_primary
+            "203.0.113.2" if secondary_has_ipv4 else None,  # public_ip
+            ["2001:db8::2"] if secondary_has_ipv6 else [],  # ipv6_addresses
+        )
+
+        # Arrange VNICs based on primary_first flag
+        if primary_first:
+            vnics_ordered = [primary_vnic, secondary_vnic]
+        else:
+            vnics_ordered = [secondary_vnic, primary_vnic]
+
+        # Setup mock VNICs
+        self.setup_mock_vnic(oci_instance, vnics_ordered)
+
+        if expect_error:
+            with pytest.raises(
+                PycloudlibError,
+                match="No public ipv4 address or ipv6 address found",
+            ):
+                _ = oci_instance.ip
+        else:
+            assert oci_instance.ip == expected_ip
```

Comment:
This can be consolidated so expect_errors value is defined in the parameterize above as  either `pytest.raises(PycloudlibError, match("No public ipv4 address or ipv6 address found")` or something like `does_not_raise()` see tests/unit_tests/lxd/test_cloud.py for the example.